상품 검색 화면 구현

hyun132 opened the pull request:
상품 검색 탭 내 화면 구현
- Navigation에 SearchResultScreen추가
- BottomNav에서 search클릭 시 SearchScreen으로 이동
- SearchScreen에서 최근검색어 / 추천검색어 표기(더미데이터)
- 검색창 클릭하거나 검색어 아이템 클릭 시 SearchResultScreen으로 이동
- 검색 결과 목록 아이템 클릭 시 상세화면으로 이동

hyun132 committed:
feature[#21] : 네비게이션에 SearchResultScreen추가, SearchScreen 파라미터 변경

hyun132 committed:
feature[#21] : SearchScreen구현. 최근/추천 검색어는 더미데이터 사용

검색어 입력 or 최근/추천 검색어 클릭해서 검색결과 화면으로 이동

hyun132 committed:
feature[#21] : SearchResultScreen,검색 기능 구현. 결과 목록 아이템 클릭 시 상세화면으로 이동

f-lab-dean commented on app/src/main/java/com/example/commerceapp/ui/SearchResultScreen.kt:
```diff
@@ -0,0 +1,147 @@
+    ) {
+
+        item {
+            Row(
```
ActionBar 성격의 아이템이 LazyColumn에 있는 것은 조금 어색할 것 같아요.
스크롤이 올라가서 탑바가 보이지 않게 되면 뒤로가기 버튼이 사라져버리니까요.
ActionBar을 상단에 고정하거나, CollapsingToolBar 같은 컴포넌트의 활용을 고려해보면 좋을 것 같습니다.

hyun132 replied:
그 생각은 못했네요! 수정해보겠습니다

f-lab-dean commented on app/src/main/java/com/example/commerceapp/ui/SearchResultScreen.kt:
```diff
@@ -0,0 +1,147 @@
+@Composable
+@Preview(showBackground = true)
+fun prevSearchResult() {
+    Row(
```
ActionBar를 별도의 Composable 함수로 만드는게 어떨까요?
동일한 코드를 복사해서 붙여넣는 형태라면, 나중에 ActionBar의 구현이 변경되었을 때 다시 코드의 복붙이 발생합니다.

hyun132 replied:
넵 별도 함수로 만들어서 반영하겠습니다!

f-lab-dean commented on app/src/main/java/com/example/commerceapp/ui/SearchResultViewModel.kt:
```diff
@@ -0,0 +1,64 @@
+    }
+
+    fun fetchProducts(keyword: String = "", page: Int = 1) {
+        if (keyword.isNullOrBlank()) return
```
코드를 부분별로 적절히 개행을 주면 가독성에 도움이 될 수 있어요.

f-lab-dean commented on app/src/main/java/com/example/commerceapp/ui/SearchScreen.kt:
```diff
@@ -0,0 +1,136 @@
+}
+
+@Composable
+fun clips(title: String, list: List<String>, onClick: (String) -> Unit) {
```
Composable 함수는 대문자로 시작하는 것이 컨벤션입니다. :)

hyun132 replied:
앗 함수 명 유의하겠습니다

f-lab-dean commented on app/src/main/java/com/example/commerceapp/ui/SearchViewModel.kt:
```diff
@@ -0,0 +1,103 @@
+        flow: Flow<ResultEntity<List<String>>>,
+        onSuccess: (List<String>) -> Unit,
+        onError: (String) -> Unit,
+        scope: CoroutineScope
```
scope 파라미터는 굳이 전달해주지 않아도 될 것 같아요.
ViewModel 내의 함수이므로 viewModelScope를 접근할 수 있기 때문입니다.

hyun132 replied:
아 그렇네요!! 수정하겠습니다

f-lab-dean commented on app/src/main/java/com/example/commerceapp/ui/SearchResultScreen.kt:
```diff
@@ -0,0 +1,147 @@
+        } else {
+            viewModel.updateKeyword(keyword)
+        }
+        viewModel.fetchProducts(keyword = keyword)
```
updateKeyword와 fetchProducts를 하나의 함수로 묶는것은 어떨까요?

hyun132 replied:
키워드가 업데이트 될 때 마다 fetchProducts를 호출하는 동작을 하는 함수를 만드는것을 말씀하시는걸까요??

f-lab-dean replied:
넵, `updateKeyword`는 UI 상태 업데이트 역할을 하고, `fetchProducts`는 키워드가 변경되었을 때 products를 업데이트 하는 역할을 가지고 있습니다.

제가 보기에 두 가지는 항상 동시에 일어나야 할 것 같습니다.
키워드가 변경되면 UI가 기억하는 keyword를 업데이트함과 동시에 products를 업데이트 하는 것 처럼 보여서요.

그래서 함수를 두 개로 나누면 keyword 업데이트 로직이 추가된다면 fetchProducts 를 또 같이 해줘야 하는 코드가 생기고, 실수할 여지가 생길 수 있어 보입니다. (물론 지금은 가능성이 매우 낮습니다.)

hyun132 replied:
아 이해했습니다! 함수 하나로 합쳐서 반영하도록 하겠습니다!
그런데 만약 모든 키워드 변경 이벤트마다 fetchProducts를 호출하는게 아니라 debounce처럼 키워드 변경 이벤트를 일부 걸러서 호출하고 싶을 때는 어떤식으로 구현하면 좋을까요?

f-lab-dean replied:
"사용자가 입력이 완료된 후 N초가 지나면 검색을 수행한다" 와 같은 요구조건으로 이해가 되는데요.
검색 기능이 있는 앱에서 일반적인 요구사항입니다.

이런 경우에는,
Rx나 Flow 같은 Stream을 활용해서 키 입력 이벤트를 debounce할 수 있습니다.
아래 게시글을 참고해서 구현하실 수 있어요.
- https://velog.io/@mraz3068/Android-Coroutine-Flow-debounce-%EB%A5%BC-%EC%9D%B4%EC%9A%A9%ED%95%9C-%EA%B2%80%EC%83%89-%EA%B8%B0%EB%8A%A5-%EA%B5%AC%ED%98%84

아마 잘 이해하고 계신 듯 한데, debounce와 throttle이 헷갈리기 쉬운 개념이어서, 다음 게시글도 한번 참고해보세요. :)
- https://medium.com/jaesung-dev/android-debounce%EC%99%80-throttle-%EC%9D%B4%ED%95%B4%ED%95%98%EA%B8%B0-e6da12d18d26

hyun132 replied:
넵 참고해서 수정해보겠습니다!
수정한 내용은 별도로pr 생성해서 올리겠습니다

hyun132 committed:
fix[#21] : [피드백 반영] appbar 별도 컴포저블로 만들어 코드 중복 제거, 코드컨벤션 적용, 앱 바 상단 고정

- SearchAppbar() 생성 및 중복 코드 제거
- appbar상단 고정
- 컴포저블 함수명 대문자로 시작하도록 수정, 파라미터 개행 추가
- 불필요한 scope파라미터 제거

f-lab-dean commented on app/src/main/java/com/example/commerceapp/ui/MainActivity.kt:
```diff
@@ -94,7 +95,7 @@ fun BottomNavigationBar(
             .fillMaxWidth()
             .padding(horizontal = 16.dp)
-            .background(Color.White),
+            .zIndex(1f),
```
zIndex는 어떤 이유로 사용하게 되었나요?

hyun132 replied:
아 이건 elevation처럼 그림자 효과를 주는 방법을 찾아보다가 실수로 남긴 부분이에요 죄송합니다.
그림자 효과 적용된 코드로 다시 커밋 올리겠습니다

f-lab-dean reviewed (commented):
코드리뷰 반영 커밋도 반영 내용에 따라 잘게 나누면 좋을 듯 합니다. :)
새로 작성된 코멘트 확인 부탁드릴게요~!

hyun132 committed:
fix[#23] : [피드백 반영] BottomNavigationBar 그림자 효과 적용

잘못 적용한 zIndex제거 및 Surface이용하여 그림자 효과 추가

hyun132 committed:
fix[#23] : [피드백 반영] 분리되어있던 검색 관련 함수 하나로 합침

검색어 변경 함수와 키워드로 검색하는 함수를 하나로 병합

f-lab-dean reviewed (approved):
키워드 입력 debounce는 별도의 PR에서 수행하셔도 좋습니다.

LGTM!
